certbot/certbot/_internal/storage.py

```diff
@@ -1243,3 +1244,42 @@ def save_successor(self, prior_version: int, new_cert: bytes, new_privkey: bytes
         self.configuration = config_with_defaults(self.configfile)
 
         return target_version
+
+    def truncate(self, current_version: int, previous_version: int,
+                 expiry_cutoff: Optional[datetime.datetime] = None) -> None:
+        """Delete unused historical certificate, chain and key items from the lineage.
+
+        A certificate version will be deleted if it is:
+          1. not the current target, and
+          2. not the previous version (current minus 1), and
+          3. expired (if expiry_cutoff is specified, otherwise all other certs are culled).
```

Comment:
I'm not sure about whether it's worthwhile to look at expiry.

Maybe it is OK to just keep current and current-1, and discard all other lineage items.

Thoughts?

Only use case that comes to mind is if somebody does `--force-renewal` a bunch of times. Since we are losing both `/{keys,csr}` and the older lineage items, the user will truly have no way of recovering those older private keys.

Reply:
Another option would be to keep around the last k, for k around like 5 or whatever. Somewhat mitigates the `--force-renewal` a bunch of times problem, and a small k is almost certainly not going to be a problem in any way that infinite k was. Expiry might be a little cleaner if we want to avoid that problem altogether, but truly I think any of these are fine.

Reply:
My initial reaction is that 5 is a lot - more than a year's worth. But you are right, it's still a good change. We also have the option to lower `k` once we are confident that this change isn't having adverse effects on DANE users or whatever.